### ISSUE TITLE
Flaky test: ConnectionCountingReturnsToZero

### ISSUE THREAD

**Issue by ryanbrandenburg** [aspnet/KestrelHttpServer#2282]
I set `ConnectionCountingReturnsToZero` to ignore for now because it [fails](http://aspnetci/viewLog.html?buildId=397052&tab=buildResultsDiv&buildTypeId=XPlat_HighSierraOsxUniverse) on OSX sometimes. It should be investigated and re-enabled once fixed.

**Comment by ryanbrandenburg** [comment]
http://aspnetci/viewType.html?buildTypeId=XPlat_Mac_MacUniverseCoreCLR&tab=buildTypeHistoryList&tag=EmptyResponse and http://aspnetci/viewType.html?buildTypeId=XPlat_HighSierraOsxUniverse&tab=buildTypeHistoryList&tag=EmptyResponse are non-exhaustive lists of builds which were affected by this.

**Comment by halter73** [comment]
I suspect this has something to do with the [Parallel.For](https://github.com/aspnet/KestrelHttpServer/blob/93b10d9060090a2080cb94731e0228427ba4a0ee/test/Kestrel.FunctionalTests/ConnectionLimitTests.cs#L166) in the test or just the fact that the test establishes 100 concurrent connections.

We could probably update the test to use less threads, but establishing a bunch of concurrent connections is fundamental to what is being tested. macOS has a [history](https://github.com/aspnet/KestrelHttpServer/blob/fdb4184dbfdbac6ee8b53a4d8d1fbfaa8f60c0c1/test/Kestrel.Transport.Libuv.Tests/LibuvTransportTests.cs#L92) of handling this badly. I know that we up the descriptor limit on our macOS agents, but that hasn't been sufficient for these types of tests when run in parallel with other tests. It's also usually not obvious that the descriptor limit is at fault for the test failure though this could probably be improved with better test logging.

The good news is that this test is *sooo* flaky on our agents (unfortunately it's still difficult to repro on my MBP since this only seems to repro when running in parallel with other tests) that it should be pretty obvious whether we've improved things once we re-enable it.

I also wonder if disabling xunit's MaxConcurrencySyncContext could help. This can easily be done on a test-by-test basis, but @anurse mentioned during standup that there was a way to do this for entire test projects. Is that the case @anurse?

**Comment by davidfowl** [comment]
We're working on making this happen across the board. I think it's the reason that kestrel tests and signalr tests and likely a bunch of other tests in our stack keep failing. See https://github.com/xunit/xunit/issues/1629

**Comment by ryanbrandenburg** [comment]
Do we have any news on this? Just shy of all of our OSX test runs fail with these.

**Comment by halter73** [comment]
I think we should skip it on mac but keep this issue open so we can fix the test and skip it post-verification.

**Comment by ryanbrandenburg** [comment]
This is not affecting a single, or even a small set of tests. If you look at a couple builds from the lists I linked above you'll see that it affects seemingly ever Kestrel functional test.

**Comment by halter73** [comment]
Disabling ConnectionCountingReturnsToZero will remove a spike in thread and connection utilization that could improve the reliability of the other tests on macOS.

**Comment by ryanbrandenburg** [comment]
That's the [one test](https://github.com/aspnet/KestrelHttpServer/commit/9c2158fd379c9c55174fff4751065434105483ae#diff-3c3a25c99aa5a7af82c985dd981ab657R136) that I already skipped before realizing that this affected basically all Kestrel functional tests, so we can count it out as a culprit.

**Comment by ryanbrandenburg** [comment]
Any movement on this?

**Comment by muratg** [comment]
Historical results: http://aspnetci/project.html?projectId=XPlat&testNameId=-8597986509470181680&tab=testDetails

@JunTaoLuo This one seems to have started failing and never passed (see the history above).  It's probably higher pri investigation than the other flaky tests.

**Comment by Tratcher** [comment]
Failures: 6 flaky failures between Jan 25 and 31 and it's been skipped since. Only Mac OSX High Sierra.

1 was a mass failure due to an API/runtime break in Memory
http://aspnetci/viewLog.html?buildId=396670&tab=buildResultsDiv&buildTypeId=XPlat_HighSierraOsxUniverse#testNameId-8597986509470181680

The other 5 all have pretty much the same stack trace:
```
Assert.Equal() Failure
          ↓ (pos 0)
Expected: HTTP/1.1 200
Actual:   
          ↑ (pos 0)
   at Microsoft.AspNetCore.Testing.TestConnection.Receive(String[] lines) in /Users/aspnetci/Documents/buildAgent/work/33bdfc1cae7b2a38/modules/KestrelHttpServer/test/shared/TestConnection.cs:line 147
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.ConnectionLimitTests.<>c__DisplayClass3_1.<<ConnectionCountingReturnsToZero>b__3>d.MoveNext() in /Users/aspnetci/Documents/buildAgent/work/33bdfc1cae7b2a38/modules/KestrelHttpServer/test/Kestrel.FunctionalTests/ConnectionLimitTests.cs:line 172
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Testing.TaskExtensions.TimeoutAfter[T](Task`1 task, TimeSpan timeout, String filePath, Int32 lineNumber)
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.ConnectionLimitTests.ConnectionCountingReturnsToZero() in /Users/aspnetci/Documents/buildAgent/work/33bdfc1cae7b2a38/modules/KestrelHttpServer/test/Kestrel.FunctionalTests/ConnectionLimitTests.cs:line 182
--- End of stack trace from previous location where exception was thrown ---
```

Various ResponseTests also failed in these runs with similar errors.

**Comment by halter73** [comment]
@Tratcher Any reason we didn't just re-enable this with a macOS skip condition?

**Comment by Tratcher** [comment]
No, feel free. The investigation didn't turn up anything and I moved on to flakier tests.

**Comment by halter73** [comment]
I'm not closing since this is still disabled on macOS, but it's now enabled everywhere else since it doesn't appear to be flaky on other OS's.

**Comment by Tratcher** [comment]
@halter73 this looks like a possible duplicate of https://github.com/aspnet/KestrelHttpServer/issues/2635

**Comment by halter73** [comment]
@Tratcher I agree about this looking like a duplicate of #2635. This particular test failing often in this way is more evidence that this issue could be caused by too many open file descriptors.
